src/storage/src/storage/bidi/connector.rs

```diff
@@ -0,0 +1,842 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::redirect::handle_redirect;
+use super::retry_redirect::RetryRedirect;
+use crate::google::storage::v2::{
+    BidiReadObjectRequest, BidiReadObjectResponse, BidiReadObjectSpec, ReadRange as ProtoRange,
+};
+use crate::read_resume_policy::{ResumeQuery, ResumeResult};
+use crate::request_options::RequestOptions;
+use crate::storage::bidi::resume_redirect::ResumeRedirect;
+use crate::storage::info::X_GOOG_API_CLIENT_HEADER;
+use crate::{Error, Result};
+use gaxi::grpc::Client as GrpcClient;
+use std::sync::{Arc, Mutex};
+use tokio::sync::mpsc::{Receiver, Sender};
+
+#[derive(Debug)]
+pub struct Connection<S = tonic::Streaming<BidiReadObjectResponse>> {
+    pub tx: Sender<BidiReadObjectRequest>,
+    pub rx: S,
+}
+
+impl<S> Connection<S> {
+    pub fn new(tx: Sender<BidiReadObjectRequest>, rx: S) -> Self {
+        Self { tx, rx }
+    }
+}
+
+/// Establishes connections to gRPC for bidi streaming reads.
+#[derive(Clone, Debug)]
+pub struct Connector<T = GrpcClient> {
+    spec: Arc<Mutex<BidiReadObjectSpec>>,
+    options: RequestOptions,
+    // This is used in testing, the client library always uses `GrpcClient`.
+    client: T,
+    reconnect_attempts: u32,
+}
+
+impl<T> Connector<T>
+where
+    T: Client + Clone + Send + 'static,
+    <T as Client>::Stream: TonicStreaming,
+{
+    pub fn new(spec: BidiReadObjectSpec, options: RequestOptions, client: T) -> Self {
+        Self {
+            spec: Arc::new(Mutex::new(spec)),
+            options,
+            client,
+            reconnect_attempts: 0_u32,
+        }
+    }
+
+    pub async fn connect(
+        &mut self,
+        ranges: Vec<ProtoRange>,
+    ) -> Result<(BidiReadObjectResponse, Connection<T::Stream>)> {
+        let throttler = self.options.retry_throttler.clone();
+        let retry = Arc::new(RetryRedirect::new(self.options.retry_policy.clone()));
+        let backoff = self.options.backoff_policy.clone();
+        let client = self.client.clone();
+        let request = BidiReadObjectRequest {
+            read_object_spec: Some((*self.spec.lock().expect("never poisoned")).clone()),
+            read_ranges: ranges,
+        };
+        let options = self.options.clone();
+        let spec = self.spec.clone();
+        let inner = async move |_| {
```

Comment:
nit: This is where the function get's interesting, everything before this is effectively. Adding a comment here would make it easier for maintainers to find the interesting part of the function.